File: Packages/NMP/Sources/NMP/Receipt.swift
// The write noun's receipt stream, pulled from `NmpReceiptStream` (#680).
// Receipts are durable FIFO facts (not disposable snapshots): live delivery
// has a finite FIFO and typed lag, while persisted outbox/redb evidence is
// replayed in deterministic pages. There is no coalescing here.

import NMPFFI

/// The ordered `WriteFact` facts a single `publish` call's write reaches
/// (guarantee #9 -- enqueue is not converged), pulled from its durable receipt
/// handle (#680). Live delivery is finite: a paused consumer that falls
/// behind receives `NMPError.factStreamLagged` and can reattach the named
/// receipt to replay the canonical persisted history. It finishes (`nil`) when
/// the engine has nothing
/// further to report for this intent (e.g. an `Ephemeral` intent may finish
/// immediately after `.sent`, a `Durable` one only after every relay has
/// reached a terminal state or given up). Iterate with `for try await`; the
/// handle is single-consumer, so a second concurrent iterator surfaces
/// `NMPError.concurrentNext` rather than hanging.
public struct ReceiptStatus: AsyncSequence, Sendable {
    public typealias Element = WriteFact

    private let handle: NmpReceiptStream
    private let iteratorGate = NMPPullIteratorGate()

    init(handle: NmpReceiptStream) {
        self.handle = handle
    }

    public func makeAsyncIterator() -> Iterator {
        let core = NMPPullIteratorCore(handle: handle, iteratorGate: iteratorGate) { status in
            WriteFact(status)
        }
        return Iterator(core: core)
    }

    public struct Iterator: AsyncIteratorProtocol {
        let core: NMPPullIteratorCore<NmpReceiptStream, WriteFact>

        public mutating func next() async throws -> WriteFact? {
            try await core.next()
        }
    }

    /// Stop delivering live status frames to this stream. The durable receipt
    /// is untouched (use `NMPEngine.cancel(receiptId:)` to cancel the write);
    /// a later `reattachReceipt` traverses the durable history. Idempotent.
    public func cancel() {
        handle.cancel()
    }
}

/// One accepted write and its live status stream. `id` is the stable
/// store-issued receipt id, usable for `reattachReceipt`/`cancel` even after
/// `status` is dropped.
public struct Receipt: Sendable {
    public let id: UInt64
    public let status: ReceiptStatus
    private let handle: NmpReceiptStream

    init(handle: NmpReceiptStream) {
        self.id = handle.id()
        self.status = ReceiptStatus(handle: handle)
        self.handle = handle
    }

    /// Await the one terminal publication answer. NMP owns receipt reduction
    /// and durable replay; callers do not fold `status` themselves.
    public func result() async throws -> ReceiptResult {
        let result = try await nmpRethrowingAsync {
            try await handle.result()
        }
        return ReceiptResult(result)
    }
}

public struct ReceiptRelayResult: Sendable, Hashable {
    public let relay: String
    public let state: RelayState

    init(_ ffi: FfiReceiptRelayResult) {
        self.relay = ffi.relay
        self.state = RelayState(ffi.state)
    }
}

/// NMP's terminal answer for one accepted write. Every known destination's
/// final state remains visible, including mixed publish/reject outcomes.
public struct ReceiptResult: Sendable, Hashable {
    public let outcome: WriteOutcome
    public let relays: [ReceiptRelayResult]

    init(_ ffi: FfiReceiptResult) {
        self.outcome = WriteOutcome(ffi.outcome)
        self.relays = ffi.relays.map(ReceiptRelayResult.init)
    }
}

public enum ReceiptReattachment: Sendable {
    case attached(Receipt)
    case notFound
    case retainedButUnreadable
}

/// Typed refusals from explicit pre-signature write cancellation.
public enum WriteCancellationOutcome: Sendable, Equatable {
    case cancelled
}

public enum NMPWriteCancellationError: Error, Sendable, Equatable {
    case unknownReceipt(receiptId: UInt64)
    case alreadySigned(receiptId: UInt64, eventId: String)
    case alreadyCompensated(receiptId: UInt64)
    case alreadySuperseded(receiptId: UInt64)
    /// The write was refused at acceptance and is already a permanently
    /// failed queue entry. There is nothing to cancel; remove it instead.
    case alreadyRefused(receiptId: UInt64)
    case persistenceFailed(receiptId: UInt64, reason: String)
    case engineClosed

    init(_ ffi: FfiCancelWriteError) {
        switch ffi {
        case .UnknownReceipt(let receiptId):
            self = .unknownReceipt(receiptId: receiptId)
        case .AlreadySigned(let receiptId, let eventId):
            self = .alreadySigned(receiptId: receiptId, eventId: eventId)
        case .AlreadyCompensated(let receiptId):
            self = .alreadyCompensated(receiptId: receiptId)
        case .AlreadySuperseded(let receiptId):
            self = .alreadySuperseded(receiptId: receiptId)
        case .AlreadyRefused(let receiptId):
            self = .alreadyRefused(receiptId: receiptId)
        case .PersistenceFailed(let receiptId, let reason):
            self = .persistenceFailed(receiptId: receiptId, reason: reason)
        case .EngineClosed:
            self = .engineClosed
        }
    }
}

/// Typed refusals from the queue-entry removal door.
public enum NMPQueueEntryRemovalError: Error, Sendable, Equatable {
    case unknownReceipt(receiptId: UInt64)
    /// The write still owns open delivery work. Cancel it first; removal is
    /// for entries nothing is going to move.
    case stillActive(receiptId: UInt64)
    case persistenceFailed(receiptId: UInt64, reason: String)
    case engineClosed

    init(_ ffi: FfiRemoveQueueEntryError) {
        switch ffi {
        case .UnknownReceipt(let receiptId):
            self = .unknownReceipt(receiptId: receiptId)
        case .StillActive(let receiptId):
            self = .stillActive(receiptId: receiptId)
        case .PersistenceFailed(let receiptId, let reason):
            self = .persistenceFailed(receiptId: receiptId, reason: reason)
        case .EngineClosed:
            self = .engineClosed
        }
    }
}

/// Typed failures from bounded publish-queue inspection.
public enum NMPPublishQueueError: Error, Sendable, Equatable {
    case invalidEventID(reason: String)
    case persistenceFailed(reason: String)
    case engineClosed

    init(_ ffi: FfiPublishQueueError) {
        switch ffi {
        case .InvalidEventId(let reason): self = .invalidEventID(reason: reason)
        case .PersistenceFailed(let reason): self = .persistenceFailed(reason: reason)
        case .EngineClosed: self = .engineClosed
        }
    }
}

extension NMPEngine {
    /// Read your own publish queue back.
    ///
    /// Answers "what have I got outstanding, and what went wrong with it"
    /// without having held a receipt stream open since acceptance. This is
    /// INSPECTION: it never blocks and never waits for settlement.
    public func publishQueue(
        afterReceiptID: UInt64? = nil,
        limit: UInt8
    ) throws -> [PublishQueueEntry] {
        do {
            return try ffi.publishQueue(
                afterReceiptId: afterReceiptID,
                limit: limit
            ).map(PublishQueueEntry.init)
        } catch let error as FfiPublishQueueError {
            throw NMPPublishQueueError(error)
        }
    }

    /// Read one bounded page of currently open obligations for a query row's
    /// exact event id. Reattach each returned receipt id to observe progress.
    public func publishQueue(
        forEventID eventID: String,
        afterReceiptID: UInt64? = nil,
        limit: UInt8
    ) throws -> [PublishQueueEntry] {
        do {
            return try ffi.publishQueueForEvent(
                eventId: eventID,
                afterReceiptId: afterReceiptID,
                limit: limit
            ).map(PublishQueueEntry.init)
        } catch let error as FfiPublishQueueError {
            throw NMPPublishQueueError(error)
        }
    }

    /// Forget one queue entry.
    ///
    /// A real TERMINATION path, not housekeeping: a write parked forever on a
    /// account whose signing provider never became available, and a permanently-failed refused entry,
    /// end no other way. A write that still owns open delivery work is
    /// refused -- cancel that one instead.
    public func removePublishQueueEntry(receiptId: UInt64) throws {
        do {
            try ffi.removePublishQueueEntry(receiptId: receiptId)
        } catch let error as FfiRemoveQueueEntryError {
            throw NMPQueueEntryRemovalError(error)
        }
    }

    /// Cancel an accepted unsigned write. Returns the durable terminal fact;
    /// repeated cancellation returns `.cancelled` idempotently.
    public func cancel(receiptId: UInt64) throws -> WriteCancellationOutcome {
        do {
            switch try ffi.cancel(receiptId: receiptId) {
            case .cancelled: return .cancelled
            }
        } catch let error as FfiCancelWriteError {
            throw NMPWriteCancellationError(error)
        }
    }

    /// Enqueue a write. Returns as soon as the intent is accepted into the
    /// outbox; `Receipt.status` streams everything that happens to it after
    /// that (M4 plan §9 -- `publish` is a one-shot enqueue call, the
    /// STREAM is where convergence is observed).
    public func publish(_ intent: WriteIntent) async throws -> Receipt {
        let handle = try nmpRethrowing {
            try ffi.publish(intent: intent.toFfi())
        }
        return Receipt(handle: handle)
    }

    /// Attach a fresh pull stream to retained receipt facts (#680): the
    /// `.attached` result carries a new `NmpReceiptStream` that transparently
    /// traverses the durable `WriteFact` history in finite pages and then
    /// streams onward. Corrupt or disappearing durable evidence is reported
    /// distinctly and never treated as absence.
    public func reattachReceipt(id: UInt64) throws -> ReceiptReattachment {
        let result = try nmpRethrowing {
            try ffi.reattachReceipt(receiptId: id)
        }
        switch result {
        case .attached(let stream):
            return .attached(Receipt(handle: stream))
        case .notFound:
            return .notFound
        case .retainedButUnreadable:
            return .retainedButUnreadable
        }
    }

}

File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/Receipt.kt
package com.nmp.sdk

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import uniffi.nmp_ffi.FfiReceiptReattachment
import uniffi.nmp_ffi.FfiCancelWriteException
import uniffi.nmp_ffi.FfiRemoveQueueEntryException
import uniffi.nmp_ffi.FfiCancelWriteOutcome
import uniffi.nmp_ffi.NmpEngineInterface
import uniffi.nmp_ffi.NmpReceiptStream

/** A stable receipt identity and its stream of retained/live write facts. */
data class Receipt(
    val id: ULong,
    val status: Flow<WriteFact>,
)

sealed interface ReceiptReattachment {
    data class Attached(val receipt: Receipt) : ReceiptReattachment

    data object NotFound : ReceiptReattachment

    data object RetainedButUnreadable : ReceiptReattachment
}

sealed class NMPWriteCancellationError(message: String) : Exception(message) {
    data class UnknownReceipt(val receiptId: ULong) :
        NMPWriteCancellationError("unknown receipt $receiptId")

    data class AlreadySigned(val receiptId: ULong, val eventId: String) :
        NMPWriteCancellationError("receipt $receiptId is already signed as $eventId")

    data class AlreadyCompensated(val receiptId: ULong) :
        NMPWriteCancellationError("receipt $receiptId is already compensated")

    data class AlreadySuperseded(val receiptId: ULong) :
        NMPWriteCancellationError("receipt $receiptId was superseded by a newer write")

    /** The write was refused at acceptance and is already a permanently
     * failed queue entry. There is nothing to cancel; remove it instead. */
    data class AlreadyRefused(val receiptId: ULong) :
        NMPWriteCancellationError("receipt $receiptId was refused at acceptance")

    data class PersistenceFailed(val receiptId: ULong, val reason: String) :
        NMPWriteCancellationError("could not persist cancellation for receipt $receiptId: $reason")

    object EngineClosed : NMPWriteCancellationError("engine already shut down")

    companion object {
        internal fun from(error: FfiCancelWriteException): NMPWriteCancellationError =
            when (error) {
                is FfiCancelWriteException.UnknownReceipt -> UnknownReceipt(error.receiptId)
                is FfiCancelWriteException.AlreadySigned ->
                    AlreadySigned(error.receiptId, error.eventId)
                is FfiCancelWriteException.AlreadyCompensated ->
                    AlreadyCompensated(error.receiptId)
                is FfiCancelWriteException.AlreadySuperseded ->
                    AlreadySuperseded(error.receiptId)
                is FfiCancelWriteException.AlreadyRefused ->
                    AlreadyRefused(error.receiptId)
                is FfiCancelWriteException.PersistenceFailed ->
                    PersistenceFailed(error.receiptId, error.reason)
                is FfiCancelWriteException.EngineClosed -> EngineClosed
            }
    }
}

enum class WriteCancellationOutcome {
    Cancelled,
}

/** Typed refusals from the queue-entry removal door. */
sealed class NMPQueueEntryRemovalError(message: String) : Exception(message) {
    data class UnknownReceipt(val receiptId: ULong) :
        NMPQueueEntryRemovalError("unknown receipt $receiptId")

    /** The write still owns open delivery work. Cancel it first; removal is
     * for entries nothing is going to move. */
    data class StillActive(val receiptId: ULong) :
        NMPQueueEntryRemovalError("receipt $receiptId still owns open delivery work; cancel it first")

    data class PersistenceFailed(val receiptId: ULong, val reason: String) :
        NMPQueueEntryRemovalError("could not remove queue entry for receipt $receiptId: $reason")

    object EngineClosed : NMPQueueEntryRemovalError("engine already shut down")

    companion object {
        internal fun from(error: FfiRemoveQueueEntryException): NMPQueueEntryRemovalError =
            when (error) {
                is FfiRemoveQueueEntryException.UnknownReceipt -> UnknownReceipt(error.receiptId)
                is FfiRemoveQueueEntryException.StillActive -> StillActive(error.receiptId)
                is FfiRemoveQueueEntryException.PersistenceFailed ->
                    PersistenceFailed(error.receiptId, error.reason)
                is FfiRemoveQueueEntryException.EngineClosed -> EngineClosed
            }
    }
}

/** Read the app's own publish queue back.
 *
 * Answers "what have I got outstanding, and what went wrong with it" without
 * having held a receipt stream open since acceptance. This is INSPECTION: it
 * never blocks and never waits for settlement. */
internal fun publishQueue(engine: NmpEngineInterface): List<PublishQueueEntry> =
    try {
        engine.publishQueue().map { PublishQueueEntry.from(it) }
    } catch (error: FfiRemoveQueueEntryException) {
        throw NMPQueueEntryRemovalError.from(error)
    }

/** Forget one queue entry.
 *
 * A real TERMINATION path, not housekeeping: a write parked forever on a
 * signer that never attached, and a permanently-failed refused entry, end no
 * other way. A write that still owns open delivery work is refused -- cancel
 * that one instead. */
internal fun removePublishQueueEntry(engine: NmpEngineInterface, receiptId: ULong) =
    try {
        engine.removePublishQueueEntry(receiptId)
    } catch (error: FfiRemoveQueueEntryException) {
        throw NMPQueueEntryRemovalError.from(error)
    }

internal fun cancelWrite(engine: NmpEngineInterface, receiptId: ULong): WriteCancellationOutcome =
    try {
        when (engine.cancel(receiptId)) {
            FfiCancelWriteOutcome.CANCELLED -> WriteCancellationOutcome.Cancelled
        }
    } catch (error: FfiCancelWriteException) {
        throw NMPWriteCancellationError.from(error)
    }

/** Build the ergonomic [Receipt] from a live [NmpReceiptStream] (#680). The
 * stable store-issued id is read synchronously off the handle; the status
 * `Flow` is a cold pull loop over `next()` -- FIFO write facts, no folding
 * and no conflation (receipts are durable facts, not disposable snapshots).
 * The live FIFO is finite and reports [NMPError.FactStreamLagged] rather than
 * growing or dropping silently; reattachment transparently traverses durable
 * facts in finite pages. Collection-scope teardown withdraws the LIVE stream
 * via `handle.cancel()`; the durable receipt itself is untouched. */
internal fun receiptFrom(stream: NmpReceiptStream): Receipt =
    Receipt(id = stream.id(), status = receiptStatusFlow(stream))

private fun receiptStatusFlow(stream: NmpReceiptStream): Flow<WriteFact> =
    flow {
        try {
            while (true) {
                val status = nmpRethrowingAsync { stream.next() } ?: break
                emit(WriteFact.from(status))
            }
        } finally {
            stream.cancel()
        }
    }

/** Enqueue immediately and retain the store-issued id needed for reattach. */
fun publishReceipt(engine: NmpEngineInterface, intent: WriteIntent): Receipt =
    receiptFrom(nmpRethrowing { engine.publish(intent.toFfi()) })

/** Map the reattachment outcome without collapsing corrupt retained
 * evidence into the same result as an unknown id (#680). Extracted with an
 * injectable `attach` so the [ReceiptReattachment.NotFound] /
 * [ReceiptReattachment.RetainedButUnreadable] distinction is unit-testable
 * without a live [NmpReceiptStream]. */
internal fun mapReceiptReattachment(
    result: FfiReceiptReattachment,
    attach: (NmpReceiptStream) -> Receipt,
): ReceiptReattachment =
    when (result) {
        is FfiReceiptReattachment.Attached -> ReceiptReattachment.Attached(attach(result.stream))
        FfiReceiptReattachment.NotFound -> ReceiptReattachment.NotFound
        FfiReceiptReattachment.RetainedButUnreadable ->
            ReceiptReattachment.RetainedButUnreadable
    }

/** Attach without collapsing corrupt retained evidence into absence. */
fun reattachReceipt(engine: NmpEngineInterface, id: ULong): ReceiptReattachment =
    mapReceiptReattachment(nmpRethrowing { engine.reattachReceipt(id) }, ::receiptFrom)

/** #591: recover a receipt after a crash that happened BEFORE the app could
 * durably persist the receipt id `publish` returned --
 * looked up by the caller's own crash-safe correlation token instead.
 * Otherwise identical to [reattachReceipt] (the by-id overload). The resolved
 * receipt id (#591) rides along on the attached stream handle itself
 * ([Receipt.id]); a token-only caller learns it there. */
fun reattachReceiptByCorrelation(engine: NmpEngineInterface, correlation: String): ReceiptReattachment =
    mapReceiptReattachment(
        nmpRethrowing { engine.reattachByCorrelation(correlation).outcome },
        ::receiptFrom,
    )
